### PR TITLE
Add tally ho

### PR DESCRIPTION
**Pull request by beemeeupnow**
# Summary

Add Tally Ho wallet as a displayed connection option. (https://tally.cash)

Previous discussion at gnosis/cowswap#2489

  # To Test

1. Run test server, access in Chrome _without_ Tally Ho installed but with another injected wallet installed (e.g. Coinbase wallet)
- [ ] Validate that 'Injected' or the tested wallet name shows up in the 'Connect to a wallet' list
2. Install Tally Ho wallet and go through onboarding in the extension. (to create or import a wallet in Tally Ho), then refresh the page (e.g. Ctrl-F5) and *leave the other wallet extension also enabled*
- [ ] Validate that "Tally Ho" shows in the wallet connection list and pops up when you try to connect
3. In the Tally Ho settings, disable Tally Ho as default and force refresh the page again
- [ ] Validate that "Tally Ho" shows in the wallet connection list, but has text to indicate that the user must enable it as default (this shows when Tally Ho is not default and another wallet occupies window.ethereum)

### CODE REVIEW

**Review by alfetopito — COMMENTED** [review]
Interesting use case

On chrome, I have [Nifty wallet](https://www.xdaichain.com/for-users/wallets/nifty-wallet) installed.
It's a MM clone and has an aggressive injection behaviour - it's always injected, even when locked.
I've installed tally, and I do see the Tally instructions to enable it
Screen Shot 2022-04-14 at 11 01 29

However, Tally is already set as default wallet but the app doesn't pick that up.

I think this is more of an issue with Nifty than anything else, but just so you know.

Even on Nifty page it mentions MM should be disabled in order for it to work. I assume the reverse is also true (you must disable Nifty to use anything else)

I'll continue testing other wallets

**Comment by alfetopito** [comment]
With Coinbase wallet installed alongside it's possible to use Tally, but doesn't behave as described/expected

When Tally is the only inject wallet, it detects itself and looks quite nice
Screen Shot 2022-04-14 at 11 08 18

When Coinbase extension is installed, the Tally setting row is shown and `Injected` button as well, even though Tally is still set as default wallet

No wallet is connected
Screen Shot 2022-04-14 at 11 15 00

Coinbase is connected
Screen Shot 2022-04-14 at 11 10 54

Tally is connected
Screen Shot 2022-04-14 at 11 11 06

**Comment by alfetopito** [comment]
With Metamask installed along side Tally, it takes precedence, even though Tally has the default wallet setting turned on
Screen Shot 2022-04-14 at 11 20 45

Can't find a way to connect Tally, same as with Nifty wallet

File: src/custom/components/WalletModal/WalletModalMod.tsx [hunk 1]
```diff
@@ -283,16 +287,46 @@ export default function WalletModal({
...
             )
+          } else if (option.name === 'Tally Ho' && tallyInstalled && !tallyIsDefault) {
+            return (
+              <Option
+                id={`connect-${key}`}
+                key={key}
+                color={'#D59B4B'}
+                header={<Trans>Tally Ho</Trans>}
+                subheader={<Trans>To use Tally Ho, enable as default in Tally Ho settings and refresh the page.</Trans>}
+                link={'https://tally.cash/'}
```
Comment (alfetopito) [inline src/custom/components/WalletModal/WalletModalMod.tsx:298]: Could this link to a help article where it instructs the user how to turn this setting on?

**Review by alfetopito — CHANGES_REQUESTED** [review]
Regardless of the use with other injected wallets, it connects and looks fine when Tally is the only wallet.
I'd be fine with, as I expect each user to take care of their extensions.

But the wallet is still not working with CowSwap's message signing.

First a few nitpicks regarding the msg signing UI:
Screen Shot 2022-04-14 at 11 27 44
1. I rather be able to expand and see/copy the full token addresses, to be able to verify them externally
2. The `appData` field (which is basically a long hex string) is going over the right side

Now when I press confirm, it fails
Screen Shot 2022-04-14 at 11 27 50

Can't see anything relevant in the console to indicate what is the reason
Screen Shot 2022-04-14 at 11 28 56

Ideally it should not fail and work out of the box :) but in case there's any special work around a wallet needs, it's handled here https://github.com/cowprotocol/cowswap/blob/abc22cb333b0631fff4de8dd0edcbe93cf8d4b35/src/custom/utils/signatures.ts#L133-L197=

I recommend you take a look and use that to understand what's missing on Tally's side

**Comment by anxolin** [comment]
> But the wallet is still not working with CowSwap's message signing.

Of course that should be a must for merging. If the user can't sign, then they can't trade. @beemeeupnow , let us know when you want us to retest this after @alfetopito 's feedback is addressed

File: src/custom/components/WalletModal/WalletModalMod.tsx (hunk 1 above)
Comment (beemeeupnow) [reply src/custom/components/WalletModal/WalletModalMod.tsx]: sounds good

**Comment by beemeeupnow** [comment]
There was an issue with the injection mechanism that was addressed in Tally Ho v0.13.1 (https://github.com/tallycash/extension/releases/tag/v0.13.1)

I will see if that covers the initial issue mentioned, then check into signing.

Thank you for the testing, I really appreciate it and will update soon!

**Comment by beemeeupnow** [comment]
synced up with current develop branch and will test further

**Comment by alfetopito** [comment]
> synced up with current develop branch and will test further
> 
> The recent releases in the Chrome store for Tally Ho have improvements to signature code and display and I will see if there's a need for anything special on this end.

Remember you can use Rinkeby to test the app and place orders to test the wallet signing ;)

**Comment by beemeeupnow** [comment]
@alfetopito Thank you for the link to the _signPayload function.

Indeed, that is where the issue lies.

Error in Tally Ho log: "Error: Attempting to sign typed data with mismatched chain IDs."

It is because domain.chainId is a string instead of a number.
To me, EIP-712 indicates clearly enough that it should be uint256 and not a string, but I understand the issue with JavaScript numbers being unable to represent such a value. Gotta love programming :smile: 

I opened tallycash/extension#1541 to see about addressing this.

**Comment by alfetopito** [comment]
> @alfetopito Thank you for the link to the _signPayload function.
> 
> Indeed, that is where the issue lies.
> 
> Error in Tally Ho log: "Error: Attempting to sign typed data with mismatched chain IDs."
> 
> It is because domain.chainId is a string instead of a number. To me, EIP-712 indicates clearly enough that it should be uint256 and not a string, but I understand the issue with JavaScript numbers being unable to represent such a value. Gotta love programming 😄
> 
> I opened [tallycash/extension#1541](https://github.com/tallycash/extension/pull/1541) to see about addressing this.

Yeah, this is exactly why we created an IntChainId signer https://github.com/cowprotocol/cowswap/blob/abc22cb333b0631fff4de8dd0edcbe93cf8d4b35/src/custom/utils/signatures.ts#L149=

We faced an issue with one version of Metamask and added it as a work around.
It has since been fixed on their side AFAICT, but the code remains.
You could add something similar to https://github.com/cowprotocol/cowswap/blob/abc22cb333b0631fff4de8dd0edcbe93cf8d4b35/src/custom/utils/signatures.ts#L27= for Tally

For additional details, see the code change where that was introduced on our side [here](https://github.com/cowprotocol/cowswap/commit/b1d817d65dc57deb299dcb049ccb2acf3529fac2) as well as the issue opened against MM's repo [here](https://github.com/MetaMask/metamask-extension/issues/11308)

**Comment by beemeeupnow** [comment]
Once Tally Ho v0.13.9 is released signing should be working
tallycash/extension#1574

**Comment by alfetopito** [comment]
> Once Tally Ho v0.13.9 is released signing should be working [tallycash/extension#1574](https://github.com/tallycash/extension/pull/1574)

Please let us know when that happens so we can give it another try

**Comment by beemeeupnow** [comment]
Tally Ho v0.13.13 has been released in the Chrome Web Store and includes the signing fixes

**Comment by alfetopito** [comment]
> Tally Ho v0.13.13 has been released in the Chrome Web Store and includes the signing fixes

Hey @beemeeupnow, doesn't seem to work yet.

Just loaded it up locally with the latest version of Tally on chrome
Screen Shot 2022-07-08 at 11 37 37

The signing modal looks very nice. One side note is that one field is going over the modal
Screen Shot 2022-07-08 at 11 36 00


But the signing ultimately fails with error `4001`
Screen Shot 2022-07-08 at 11 37 50

**Comment by alfetopito** [comment]
https://user-images.githubusercontent.com/43217/177980678-38ea2d58-9850-459e-ae41-b7723607ec54.mov

**Comment by anxolin** [comment]
That's strange, it worked for me in 
image

Should we detect and try to personalise a bit adding the logo of Tally if it's the connected wallet?
image

Also, I wouldn't add the option for tally, but if we know the injected provider is Tally, maybe we can personalise the "Injected option"?

I believe we can test `ethereum.isTally`

image

**Comment by elena-zh** [comment]
I have run all types of transactions on Staging/Barn with Tally web extension (after I disabled MM and Coinbase apps), and everything worked fine. 
 The only thing is that the wallet is marked as 'Injected' (Anxo has described it above), but this issue is related to other injected wallets and will be fixed in another PR

**Comment by alfetopito** [comment]
I tried in prod like Anxo and Elena, and it still failed to me.
I assume then it's something with my local setup.
I'll reset my account and try again.

**Comment by alfetopito** [comment]
After resetting Tally (removing and installing it again) it worked!
Both on production and on this PR.

Then, it might be ready to merge

@anxolin any concerns?

**Comment by Naxsun** [comment]
Hey everyone. Was wondering what the next step would be here, and if there's something we can do from our side

**Comment by alfetopito** [comment]
> Hey everyone. Was wondering what the next step would be here, and if there's something we can do from our side

Hey @Naxsun, we are in the process of re-working how the wallets are handled, so this might change slightly.
Still, with the changes made by @beemeeupnow to Tally upstream, it's now fully compatible with Cowswap and can already be used in the current prod version 🥳 

Will keep this on hold for now.

**Comment by Naxsun** [comment]
@alfetopito got it, thank you for the update 👍

**Pull request closed without merging**